index.ts

```diff
@@ -7,28 +7,167 @@ if (MATBAS_BUILD === 'debug') {
 }
 
 import { getTracer } from './debug';
-
-import { container } from './container';
-import { Argv, cli, Env } from './cli';
+import { container, Container } from './container';
 import { Config } from './config';
-import { Host } from './host';
+import { BaseException } from './exceptions';
+import { BaseFault, RuntimeFault, UsageFault } from './faults';
+import { ConsoleHost, Host } from './host';
+import { Exit, ExitCode } from './exit';
 
 const tracer = getTracer('main');
 
-function parseArgs(argv: Argv, env: Env): Config {
-  return Config.load(argv, env);
+/**
+ * Command line arguments.
+ */
+export type Argv = typeof process.argv;
+
+/**
+ * Command line environment.
+ */
+export type Env = typeof process.env;
+
+/**
+ * Options for running a CLI.
+ */
+export interface RuntimeOptions {
+  /**
+   * Command line arguments.
+   */
+  argv: Argv;
+
+  /**
+   * Command line environment.
+   */
+  env: Env;
+}
+
+//
+// Run a script.
+//
+async function script<H extends Host>(
+  { executor }: Container<H>,
+  filename: string,
+) {
+  await executor.using(async () => {
+    await executor.load(filename);
+    await executor.run();
+  });
+}
+
+//
+// Run the REPL.
+//
+async function repl<H extends Host>({ executor, host }: Container<H>) {
+  await executor.using(async () => {
+    while (true) {
+      try {
+        const input = await executor.prompt();
+        await executor.eval(input);
+      } catch (err) {
+        // TODO: This sort of logic is duplicated in the executor.
+        // On that note, the logic in the executor is probably buggy.
+        if (err instanceof BaseFault || err instanceof Exit) {
+          throw err;
+        }
+
+        if (err instanceof BaseException) {
+          host.writeException(err);
+        }
+
+        throw RuntimeFault.fromError(err, null);
+      }
+    }
+  });
 }
 
-async function run(config: Config, host: Host): Promise<void> {
-  await tracer.span('main', async () => {
-    const { translator } = container(config, host);
+/**
+ * The main entry point for the application.
+ *
+ * @param options Runtime options.
+ */
+export async function main({ argv, env }: RuntimeOptions): Promise<void> {
+  const host: Host = new ConsoleHost();
```

Comment:
Overriding the Host will probably remain useful - it will let me run programs from within tests and assert the output on a MockConsoleHost.